week-4/E-arrays-of-objects/exercise-2.js

```diff
@@ -7,50 +7,84 @@ Each destination has a name, a distance from Glasgow, and a list of transportati
 3) Print in the console all the destination names more than 300 kms far away and reachable by train.
 */
 
-
 var destination1 = {
-    destinationName: "Edinburgh",
-    distanceKms: 80,
-    transportations: ["car", "bus", "train"]
+  destinationName: "Edinburgh",
+  distanceKms: 80,
+  transportations: ["car", "bus", "train"]
 };
 
 var destination2 = {
-    destinationName: "London",
-    distanceKms: 650,
-    transportations: ["car", "bus", "train"]
+  destinationName: "London",
+  distanceKms: 650,
+  transportations: ["car", "bus", "train"]
 };
 
 var destination3 = {
-    destinationName: "Paris",
-    distanceKms: 900,
-    transportations: ["train", "plane"]
+  destinationName: "Paris",
+  distanceKms: 900,
+  transportations: ["train", "plane"]
 };
 
 var destination4 = {
-    destinationName: "Dublin",
-    distanceKms: 350,
-    transportations: ["plane", "ferry"]
+  destinationName: "Dublin",
+  distanceKms: 350,
+  transportations: ["plane", "ferry"]
 };
 
-var travelDestinations = [destination1, destination2, destination3, destination4];
+var travelDestinations = [
+  destination1,
+  destination2,
+  destination3,
+  destination4
+];
 
 /*
 DO NOT EDIT ANYTHING ABOVE THIS LINE
 WRITE YOUR CODE BELOW
 */
+function travelDestinationsLessThan500Kms(destination) {
+  if (destination.distanceKms <= 500) {
+    return destination.destinationName;
+  }
+}
+function destinationReachByFerry(destination) {
+  if (destination.transportations.includes("ferry")) {
+    return destination.destinationName;
+  }
+}
+function DestNameMoreThan300KmsByTrain(destination) {
+  if (
+    destination.distanceKms > 300 &&
+    destination.transportations.includes("train")
+  ) {
+    return destination.destinationName;
+  }
+}
+var destinationNamesWithin500Kms = travelDestinations
+  .filter(travelDestinationsLessThan500Kms)
+  .map(travelDestinationsLessThan500Kms); // Complete here
 
+var destinationNameReachableByFerry = travelDestinations
+  .filter(destinationReachByFerry)
+  .map(destinationReachByFerry); // Complete here
 
-var destinationNamesWithin500Kms = // Complete here
-
-var destinationNameReachableByFerry = // Complete here
-
-var destinationNamesMoreThan300KmsAwayByTrain = // Complete here (PRINT THE RESULT IN THE CONSOLE USING FOREACH)
-
+var destinationNamesMoreThan300KmsAwayByTrain = travelDestinations
+  .filter(DestNameMoreThan300KmsByTrain)
+  .map(DestNameMoreThan300KmsByTrain); // Complete here (PRINT THE RESULT IN THE CONSOLE USING FOREACH)
```

Comment:
This works, but possibly not in a way that you would expect.
`travelDestinationsLessThan500Kms` will return `undefined` if the distance is under 500. This is close enough to `false` for filter to work, so it filters it. You then call the same function again with `map`.
Similar logic applies to the others.

A better way of doing this is to have two functions for each - one that filters and one that maps e.g.
```
var destinationNamesWithin500Kms = 
    travelDestinations.filter(
        place => place.distanceKms < 500
    ).map(
        place => place.destinationName
    );
```